### PR TITLE
fix: increment query_id even on failed queries to prevent id offset during restore

### PR DESCRIPTION
**Pull request by stevenpyzhang**
### Description 
Issue: https://github.com/confluentinc/ksql/issues/3269

To the best of my understanding, the command topic doesn't store statements that don't succeed while the server is running, so CREATE statements or any statement that failed while the server is running won't be executed during restore and therefore won't affect the QueryId.

Sample of behavior currently

Before every scenario run these
```
CREATE STREAM foo(age BIGINT) WITH (KAFKA_TOPIC='foo',  VALUE_FORMAT='JSON');
CREATE STREAM bar (age BIGINT) WITH (KAFKA_TOPIC='bar', VALUE_FORMAT='JSON');
CREATE STREAM test(age BIGINT) WITH (KAFKA_TOPIC='test',  VALUE_FORMAT='JSON');
```
_______ 
```
CREATE STREAM foo_copy AS SELECT * from foo; 
TERMINATE CSAS_FOO_COPY_0;

CREATE STREAM bar_copy AS SELECT * from bar; 
TERMINATE CSAS_BAR_COPY_1;
```
No queries remaining
Close server, Delete topic foo, reboot server
CSAS_BAR_COPY_0 is running 
_________________________________________________________________________________
```
INSERT INTO test  SELECT age FROM foo;
TERMINATE InsertQuery_0;

INSERT INTO test  SELECT age FROM bar;
TERMINATE InsertQuery_1;
```
No queries remaining
Close server, Delete topic foo, reboot server
InsertQuery_0 (INSERT INTO test  SELECT age FROM bar) is running
___________________________________________________________________________________
```
INSERT INTO foo SELECT age FROM test;
TERMINATE InsertQuery_0;

INSERT INTO bar SELECT age FROM test;
TERMINATE InsertQuery_1;
```
No queries remaining
Close server, Delete topic foo, reboot server
InsertQuery_0 (INSERT INTO bar  SELECT age FROM test) is running
Logs indicate a InvalidColumnReferenceException thrown when trying to getSource of InsertInto node for inserting into stream foo (different exception to first two examples)

### Testing done 
Unit tests 
Local testing:
```
CREATE STREAM test(age BIGINT) WITH (KAFKA_TOPIC='test',  VALUE_FORMAT='JSON');
CREATE STREAM foo(age BIGINT) WITH (KAFKA_TOPIC='foo',  VALUE_FORMAT='JSON');
CREATE STREAM foo_copy AS SELECT * from foo; 
INSERT INTO test  SELECT age FROM foo;
INSERT INTO foo SELECT age FROM test;
TERMINATE CSAS_FOO_COPY_0;
TERMINATE InsertQuery_1;
TERMINATE InsertQuery_2;

CREATE STREAM bar (age BIGINT) WITH (KAFKA_TOPIC='bar', VALUE_FORMAT='JSON');
CREATE STREAM bar_copy AS SELECT * from bar; 
INSERT INTO test  SELECT age FROM bar;
INSERT INTO bar  SELECT age FROM test;
TERMINATE CSAS_BAR_COPY_3;
TERMINATE InsertQuery_4;
TERMINATE InsertQuery_5;

# Behavior on master
SHOW QUERIES;
 Query ID        | Kafka Topic | Query String                                                                               
----------------------------------------------------------------------------------------------------------------------------
 CSAS_BAR_COPY_0 | BAR_COPY    | CREATE STREAM BAR_COPY WITH (KAFKA_TOPIC='BAR_COPY', PARTITIONS=1, REPLICAS=1) AS SELECT *
FROM BAR BAR; 
 InsertQuery_1   | TEST        | INSERT INTO test  SELECT age FROM bar;                                                     
 InsertQuery_2   | BAR         | INSERT INTO bar  SELECT age FROM test;                                                     
----------------------------------------------------------------------------------------------------------------------------

# Behavior after change
SHOW QUERIES;
----------------------------------------------------------------------------------------------------------------------------                                                  
----------------------------------------------------------------------------------------------------------------------------
```
### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

**Comment by big-andy-coates** [comment]
I'm with @almog on this. Now that we validate statements by executing them against a sandboxed clone of the engine that has been updated to a _known_ offset we can assign query ids and bake them into the statement written to the command topic. 

This doesn't fix issues with any statement already in an installations command topic from a previous version of KSQL, but going forward I think this is OK.  Better to have a good solid fix going forward that doesn't fix historic issues, than a less solid fix that fixes historic issues.

Also, we'll need a solution for standalone mode, i.e. when there is no command topic.  Like interactive mode, queries restarting after a server bounce should restart with the same query id.

I'm not sure of the best solution for assigning query ids in standalone mode. A simple one would be to assign ids based on the order of queries in the sql file. Though we may be able to come up with a better one might allow some editing of the sql file without changing query ids.  However, that's probably out of scope for this PR.

Basically, we'll need a way of injecting a query id supplier into the engine when its building queries. This will be needed in the Rest server when validating queries by building them in a sandbox, and will be needed by standalone.  This supplier should only be used if the statement doesn't have the queryid baked in, e.g. `WITH(QUERY_ID='x')`

Question: should we allow users to specify the query id in the with clause, or fail if that is supplied?

**Comment by rodesai** [comment]
+1 to @almog's suggestion. The offset would be the right thing to use here. However, to avoid query id collisions this would also have to be done in conjunction with a change to only execute statements whose offsets are the same as the offset they were validated against. I think this is the right thing to do for other reasons as well, which we've discussed in the past here: https://github.com/confluentinc/ksql/issues/2435. Furthermore, if we do this I think we should also change our error handling to ensure we never skip over statements whose actual offset match their validation offset. But, that's out of scope for fixing this particular issue.

**Comment by agavra** [comment]
> However, to avoid query id collisions this would also have to be done in conjunction with a change to only execute statements whose offsets are the same as the offset they were validated against.

@rodesai I think we can have one without the other. The validation will be done anyway before we execute it - worst case we "wasted" a queryID - unless I'm missing something!

**Comment by rodesai** [comment]
> worst case we "wasted" a queryID - unless I'm missing something!

I think the worst case is that we share a query ID for different queries. Imagine 2 thread/servers A and B are both processing an INSERT_INTO, and the current offset is O. Both will choose a query ID O+1, and enqueue that onto the command topic. Unless the executor has some way to just choose 1 of the 2 to run, we'll wind up with 2 queries using the same query id (and consequently consumer group, state stores, etc)

**Comment by agavra** [comment]
ah I had imagined using the offset from kafka at the time of enqueuing (perhaps a different one than the one we chose to validate against) that might not be possible though...

**Comment by rodesai** [comment]
Ohhh you mean just use the offset of the record observed when you consume? That would work.

**Comment by rodesai** [comment]
As discussed, I think we'll need this change even if we change how we generate query IDs to handle backwards-compatibility - old commands won't have query ids (or won't have ids based on offsets), and we should try to preserve those.

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/StatementExecutor.java [hunk 1]
```diff
@@ -196,6 +198,11 @@ private void handleStatementWithTerminatedQueries(
       executeStatement(
           statement, command, commandId, commandStatusFuture, mode);
     } catch (final KsqlException exception) {
+      final Throwable rootCause = ExceptionUtils.getRootCause(exception);
+      if (mode == Mode.RESTORE && rootCause instanceof KsqlMissingSourceException) {
```
Comment (vcrfxia) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/StatementExecutor.java]: Apologies if I've missed something as I'm late to the game here: there are multiple reasons a statement could during recovery, besides a missing source, so I don't believe this approach works. (For example, I originally discovered the recovery bug when a statement failed to execute because Schema Registry rejected a sink schema due to a failed schema compatibility check.)

If we're pursuing this fix for legacy query IDs (rather than updating how query IDs are generated and leaving the bug for legacy query IDs, as proposed in the discussion above), then we should always increment the query ID on recovery, regardless of the reason for failure.

**Comment by stevenpyzhang** [comment]
Moving this discussion/change to https://github.com/confluentinc/ksql/pull/3296
